nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestSQLRepositoryAPI.java

```diff
@@ -4431,26 +4437,168 @@ public void testLockingBeforeSave() {
+        assertEquals(CoreSession.RETAIN_UNTIL_INDETERMINATE, session.getRetainUntil(doc.getRef()));
+        doc.refresh();
+        assertEquals(CoreSession.RETAIN_UNTIL_INDETERMINATE, doc.getRetainUntil());
+
+        // cannot be deleted
+        checkUndeletable(folder, doc);
+        // bob cannot modify it
+        try (CloseableCoreSession bobSession = openSessionAs("bob")) {
+            assertFalse(bobSession.hasPermission(doc.getRef(), SecurityConstants.REMOVE));
+        }
+
+        // then reduce retention to an actual date in the future
+        Calendar retainUntil = Calendar.getInstance();
+        retainUntil.add(Calendar.HOUR, 1);
+        session.setRetainUntil(doc.getRef(), retainUntil, null);
+        assertEquals(retainUntil, session.getRetainUntil(doc.getRef()));
+        doc.refresh();
+        assertEquals(retainUntil, doc.getRetainUntil());
+
+        // still available when freshly detached
+        doc = session.getDocument(doc.getRef());
+        doc.detach(true);
+        assertEquals(retainUntil, doc.getRetainUntil());
+        doc.attach(session.getSessionId());
+
+        // cannot be deleted
+        checkUndeletable(folder, doc);
+        // bob cannot modify it
+        try (CloseableCoreSession bobSession = openSessionAs("bob")) {
+            assertFalse(bobSession.hasPermission(doc.getRef(), SecurityConstants.REMOVE));
+        }
+
+        // cannot reduce retention
+        Calendar retainUntilEarlier = Calendar.getInstance();
+        retainUntilEarlier.add(Calendar.MINUTE, 5);
+        try {
+            session.setRetainUntil(doc.getRef(), retainUntilEarlier, null);
+            fail("Should fail because cannot reduce retention time");
+        } catch (PropertyException e) {
+            assertTrue(e.getMessage(), e.getMessage().startsWith("Cannot reduce retention time"));
+        }
+    }
+
+    @Test
+    public void testLegalHold() {
+        DocumentModel folder = session.createDocumentModel("/", "fold", "Folder");
+        folder = session.createDocument(folder);
+        ACP acp = new ACPImpl();
+        ACL acl = new ACLImpl();
+        acl.add(new ACE("bob", "Remove", true));
+        acp.addACL(acl);
+        folder.setACP(acp, true);
+        DocumentModel subFolder = session.createDocumentModel("/fold", "subfold", "Folder");
+        session.createDocument(subFolder);
+        DocumentModel doc = session.createDocumentModel("/fold/subfold", "doc", "File");
+        doc = session.createDocument(doc);
+        assertFalse(doc.hasLegalHold());
+
+        // must be a record
+        try {
+            session.setRetainUntil(doc.getRef(), null, null);
+            fail("Should fail because document is not a record");
+        } catch (PropertyException e) {
+            assertEquals("Document is not a record", e.getMessage());
+        }
+
+        // make it a record
+        session.makeRecord(doc.getRef());
+        // bob can modify it
```

Comment:
can we modify the comment `bob can modify it` because there is no modification her, the fact that the `bob` user didn't have permission is related to `RetentionAndHoldSecurityPolicy`, wdyt if we replace the comment by something `bob had not Remove permission because of RetentionAndHoldSecurityPolicy` :)

Reply:
Changing to "bob can delete" it which is what we're actually testing (permission and actual method). The fact that it's implemented through a security policy is a detail for the test.